coding/trie_reader.hpp

```diff
@@ -9,36 +9,29 @@
 
 namespace trie
 {
-template <class TValueReader>
-class LeafIterator0 : public Iterator<typename TValueReader::ValueType>
+template <class TValueList, typename TSerializer>
+class LeafIterator0 : public Iterator<TValueList>
 {
 public:
-  using ValueType = typename TValueReader::ValueType;
+  using TValue = typename TValueList::TValue;
+  using Iterator<TValueList>::m_valueList;
 
   template <class TReader>
-  LeafIterator0(TReader const & reader, TValueReader const & valueReader)
+  LeafIterator0(TReader const & reader, TSerializer const & serializer)
   {
-    uint32_t const size = static_cast<uint32_t>(reader.Size());
     ReaderSource<TReader> src(reader);
-    while (src.Pos() < size)
-    {
-      this->m_value.push_back(ValueType());
-#ifdef DEBUG
-      uint64_t const pos = src.Pos();
-#endif
-      valueReader(src, this->m_value.back());
-      ASSERT_NOT_EQUAL(pos, src.Pos(), ());
-    }
-    ASSERT_EQUAL(size, src.Pos(), ());
+    if (src.Size() > 0)
+      m_valueList.Deserialize(src, 1 /* valueCount */, serializer);
```

Comment:
This is true iff TValueList is a CBV. Probably you need two individual specializations of LeafIterator0: the first one for old value list and the second one for CBV.